core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala

```diff
@@ -41,6 +42,12 @@ private[storage] class BlockManagerDecommissioner(
   private val maxReplicationFailuresForDecommission =
     conf.get(config.STORAGE_DECOMMISSION_MAX_REPLICATION_FAILURE_PER_BLOCK)
 
+  // Used for tracking if our migrations are complete.
+  @volatile private var lastRDDMigrationTime: Long = 0
+  @volatile private var lastShuffleMigrationTime: Long = 0
+  @volatile private var rddBlocksLeft: Boolean = true
+  @volatile private var shuffleBlocksLeft: Boolean = true
+
```

Comment:
Should we fold all these into a separate class with mutable fields such that we can simply return a clone of it when `lastMigrationInfo` ?

I am not sure if it is worth the trouble, since it is only for info logging: In case we notice an executor that is not going away.

Reply:
We use this to determine if the blocks are migrated and these can be changed from seperate threads so we'd have to switch to synchronized.